% https://github.com/uio-latex/phduio-article-based
% localdefs

\newcommand{\ctext}[3][RGB]{%
  \begingroup
  \definecolor{hlcolor}{#1}{#2}\sethlcolor{hlcolor}%
  \hl{#3}%
  \endgroup
}


%end local defs

\documentclass[colophon, english]{phduio}

\usepackage{phdstyle}   % Custom style
\usepackage{kantlipsum} % Dummy text

%\makenoidxglossaries
\makeglossaries
\newacronym{AMQ}{AMQ}{Approximate Memmbership Query}
\newacronym{API}{API}{Application Programming Interface}
\newacronym{BQF}{BQF}{Backpack Quotient Filter}
\newacronym{cAMQ}{cAMQ}{counting Approximate Memmbership Query}
\newacronym{CBF}{CBF}{Counting Bloom Filter}
\newacronym{cli}{cli}{Command Line Interface}
\newacronym{CNV}{CNV}{Copy Number Variant}
\newacronym{CQF}{CQF}{Counting Quotient Filter}
\newacronym{DNA}{DNA}{Deoxyribonucleic acid}
\newacronym{dsDNA}{dsDNA}{double stranded DNA}
\newacronym{ENA}{ENA}{European Nucleotide Archive}
\newacronym{GWAS}{GWAS}{Genome Wide Association Study}
\newacronym{HiFi}{HiFi}{High Fidelity}
\newacronym{HOR}{HOR}{Higher Order Repeat}
\newacronym{HPRC}{HPRC}{Human Pangenome Reference Consortium}
\newacronym{kff}{kff}{\kmer file format}
\newacronym{MPHF}{MPHF}{Minimal Perfect Hash Functions}
\newacronym{MSA}{MSA}{Multiple Sequence Alignment}
\newacronym{NIH}{NIH}{National Institute of Health}
\newacronym{NGS}{NGS}{Next Generation Sequencing}
\newacronym{ONT}{ONT}{Oxford Nanopore Technologies}
\newacronym{PCR}{PCR}{Polymerase Chain Reaction}
\newacronym{QC}{QC}{Quality Control}
\newacronym{QF}{QF}{Quotient Filter}
\newacronym{RNA}{RNA}{Ribonucleic acid}
\newacronym{RSQF}{RSQF}{Rank Select Quotient Filter}
\newacronym{SIMD}{SIMD}{Single Instruction Multiple Data}
\newacronym{SNP}{SNP}{Single Nucleotide Polymorphism}
\newacronym{SNV}{SNV}{Single Nucleotide Variant}
\newacronym{SRA}{SRA}{Sequence Read Archive}
\newacronym{ssDNA}{ssDNA}{single stranded DNA}
\newacronym{SV}{SV}{Structural Variation}
\newacronym{TGS}{TGS}{Third Generation Sequencing}
\newacronym{T2T}{T2T}{Telomere to Telomere consortium}

%HbS,DARC,STARR-seq, ATAC-seq, ChIP-seq, RNA-seq

\author{Francesco Andreace}
\title{Thesis Manuscript Draft}
\subtitle{Analysis of human pangenome graphs using \kmer based applications}
\uni{Sorbonne Université}
\institute{Department of Computational Biology - Insitut Pasteur Paris, Université de Paris Cité}
\affiliation{Ecole Doctorale Informatique, Télécommunications et Electronique EDITE
	(ED130)}


\includeonly
{
	sections/dedication,
	sections/preface,
	sections/summary,
	sections/introduction,
	sections/background,
	sections/pushing_the_limit,
	sections/reducing_complexity,
	sections/explorations,
	sections/perspectives,
	sections/papers,
	sections/conclusions
}

%	\frontmatter
\begin{document}
	 \frontmatter
	 
	\maintitle
	
	\include{sections/dedication}
	\include{sections/preface}
	\include{sections/summary}
	
	\cleartorecto
	\microtypesetup{protrusion = false}
	\tableofcontents    % Or \tableofcontents*
	\cleartorecto
	\listoffigures      % Or \listoffigures*
	\cleartorecto
	\listoftables       % Or \listoftables*
	\microtypesetup{protrusion = true}
	
	\mainmatter
	\include{sections/introduction}
	\include{sections/background}
	\include{sections/pushing_the_limit}
	\include{sections/reducing_complexity}
	\include{sections/papers}
	\include{sections/perspectives}
	\include{sections/conclusions}
	
	%\printnoidxglossary[type=\acronymtype,title=Acronyms]
	\printglossary[type=\acronymtype]
	
\end{document}