 \chapter{Summary}
\label{sec:Summary}
\section{English summary}
The human reference genome, which has served as the foundation of genomics for over 20 years, contains gaps and computationally simulated sequences, limiting its usefulness. The Telomere-to-Telomere consortium recently completed the first full sequence of a haploid human genome (T2T-CHM13), improving understanding of genomic variations. However, a single genome cannot represent the full genetic diversity of the human population, as structural variants are often missed. To address reference bias, a shift toward pangenomes, which are models that capture multiple alleles from a collection of genomes, is being pursued. These pangenomes can be represented using graph-based models, and this work contributes to computational pangenomics by analyzing different tools for building such graphs and by proposing new data structures for efficiently handling \kmer sets, which offer advantages for scalability, speed, and flexibility in genomic analysis.

This dissertation focuses on computational pangenomics and proposes two main contributions. The first is an analysis of the largest human pangenome constructed at the time, comparing pangenome graph representations using five advanced tools. The analysis highlights differences in how these tools capture genetic variations, assisting users in selecting the best graph type for their needs. The second contribution involves improving data structures for representing \kmer sets and their representation of metadata. \kmer-based data structures offer advantages for genomic analysis in terms of scalability, speed, and memory efficiency, and the dissertation focuses on my contribution on three new data structures that advance the current state of \kmer representations.

\section{Résumé en français}